File: frontend/types/analytics.ts
export type AnalyticsIntent =
  | "sales_summary"
  | "sales_trend"
  | "top_products"
  | "top_customers"
  | "branch_performance"
  | "inventory_summary"
  | "payment_summary"
  | "invoice_summary"
  | "supplier_summary"
  | "procurement_summary";

export type AnalyticsPeriod = "daily" | "weekly" | "monthly";

export type AnalyticsFilters = {
  date_from?: string;
  date_to?: string;
  branch_id?: string;
  limit?: number;
  period?: AnalyticsPeriod;
};

export type AnalyticsResponse<T> = {
  intent: AnalyticsIntent;
  date_from: string;
  date_to: string;
  data: T;
};

export type SalesSummary = {
  total_sales: string | number;
  order_count: number;
  average_order_value: string | number;
};

export type SalesTrend = {
  breakdown: Array<{
    period: string;
    sales: string | number;
    order_count: number;
  }>;
};

export type TopProducts = {
  products: Array<{
    product_id: string;
    sku: string;
    name: string;
    quantity: string | number;
    revenue: string | number;
  }>;
};

export type TopCustomers = {
  customers: Array<{
    customer_id: string;
    name: string;
    email: string | null;
    order_count: number;
    revenue: string | number;
  }>;
};

export type BranchPerformance = {
  branches: Array<{
    branch_id: string;
    code: string;
    name: string;
    order_count: number;
    sales: string | number;
  }>;
};

export type InventorySummary = {
  low_stock: InventoryRow[];
  out_of_stock: InventoryRow[];
  inventory_value: string | number;
};

export type InventoryRow = {
  product_id: string;
  sku: string;
  name: string;
  branch_id: string;
  branch_name: string;
  quantity: string | number;
  reorder_level: string | number;
};

export type PaymentSummary = {
  total: string | number;
  by_method: Array<{ method: string; total: string | number }>;
  by_status: Array<{ status: string; total: string | number }>;
};

export type InvoiceSummary = {
  total: string | number;
  invoice_count: number;
  status_counts: Record<string, number>;
  overdue_count: number;
};

export type SupplierSummary = {
  suppliers: Array<{
    supplier_id: string;
    supplier_name: string;
    product_count: number;
    inventory_value: string | number;
  }>;
};

export type ProcurementSummary = {
  purchase_order_count: number;
  pending_approvals: number;
  pending_receipts: number;
  purchasing_total: string | number;
  status_counts: Record<string, number>;
  top_suppliers: Array<{ supplier_id: string; supplier_name: string; purchase_order_count: number; purchasing_total: string | number }>;
};

export type AnalyticsDashboardData = {
  summary: AnalyticsResponse<SalesSummary>;
  trend: AnalyticsResponse<SalesTrend>;
  products: AnalyticsResponse<TopProducts>;
  customers: AnalyticsResponse<TopCustomers>;
  branches: AnalyticsResponse<BranchPerformance>;
  inventory: AnalyticsResponse<InventorySummary>;
  payments: AnalyticsResponse<PaymentSummary>;
  invoices: AnalyticsResponse<InvoiceSummary>;
  suppliers: AnalyticsResponse<SupplierSummary>;
  procurement: AnalyticsResponse<ProcurementSummary>;
};

export type NaturalLanguageResponse = {
  supported: boolean;
  message: string;
  intent: AnalyticsIntent | null;
  date_from: string | null;
  date_to: string | null;
  data: Record<string, unknown> | null;
};

File: backend/app/api/v1/routes/analytics.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.encoders import jsonable_encoder
from pydantic import ValidationError
from sqlalchemy.orm import Session

from app.core.security import require_role
from app.db.session import get_db
from app.schemas.analytics import (
    AnalyticsFilters,
    AnalyticsQueryRequest,
    AnalyticsResponse,
    NaturalLanguageQueryRequest,
    NaturalLanguageQueryResponse,
)
from app.services import analytics as analytics_service
from app.services.audit import log_activity

router = APIRouter(prefix="/analytics", tags=["analytics"])


def _execute(
    payload: AnalyticsQueryRequest,
    membership: Any,
    db: Session,
) -> AnalyticsResponse:
    try:
        start, end, data = analytics_service.execute(
            db,
            membership.organization_id,
            payload.intent,
            payload,
        )
        log_activity(
            db,
            organization_id=membership.organization_id,
            user=membership.user,
            action="ANALYTICS_QUERY_EXECUTED",
            entity_type="ANALYTICS",
            details=f"Executed controlled analytics intent: {payload.intent}",
        )
        db.commit()
        return AnalyticsResponse(intent=payload.intent, date_from=start, date_to=end, data=data)
    except Exception as exc:
        db.rollback()
        if isinstance(exc, ValueError | KeyError):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Analytics request could not be completed") from exc


def _filters(
    date_from: datetime | None = Query(None),
    date_to: datetime | None = Query(None),
    branch_id: UUID | None = Query(None),
    limit: int = Query(10, ge=1, le=50),
    period: str = Query("daily", pattern="^(daily|weekly|monthly)$"),
) -> AnalyticsFilters:
    try:
        return AnalyticsFilters(date_from=date_from, date_to=date_to, branch_id=branch_id, limit=limit, period=period)
    except ValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=jsonable_encoder(exc.errors()),
        ) from exc


@router.get("/sales-summary", response_model=AnalyticsResponse)
def sales_summary(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="sales_summary", **filters.model_dump()), membership, db)


@router.get("/sales-trend", response_model=AnalyticsResponse)
def sales_trend(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="sales_trend", **filters.model_dump()), membership, db)


@router.get("/top-products", response_model=AnalyticsResponse)
def top_products(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="top_products", **filters.model_dump()), membership, db)


@router.get("/top-customers", response_model=AnalyticsResponse)
def top_customers(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="top_customers", **filters.model_dump()), membership, db)


@router.get("/branches", response_model=AnalyticsResponse)
def branches(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="branch_performance", **filters.model_dump()), membership, db)


@router.get("/inventory", response_model=AnalyticsResponse)
def inventory(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="inventory_summary", **filters.model_dump()), membership, db)


@router.get("/payments", response_model=AnalyticsResponse)
def payments(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="payment_summary", **filters.model_dump()), membership, db)


@router.get("/invoices", response_model=AnalyticsResponse)
def invoices(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="invoice_summary", **filters.model_dump()), membership, db)


@router.get("/suppliers", response_model=AnalyticsResponse)
def suppliers(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="supplier_summary", **filters.model_dump()), membership, db)


@router.get("/procurement", response_model=AnalyticsResponse)
def procurement(
    filters: AnalyticsFilters = Depends(_filters),
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    return _execute(AnalyticsQueryRequest(intent="procurement_summary", **filters.model_dump()), membership, db)


@router.post("/query", response_model=NaturalLanguageQueryResponse)
def natural_language_query(
    payload: NaturalLanguageQueryRequest,
    membership=Depends(require_role(["admin", "manager"])),
    db: Session = Depends(get_db),
) -> Any:
    intent = analytics_service.parse_question(payload.question)
    if intent is None:
        log_activity(
            db,
            organization_id=membership.organization_id,
            user=membership.user,
            action="ANALYTICS_QUERY_UNSUPPORTED",
            entity_type="ANALYTICS",
            details="Received unsupported natural-language analytics question",
        )
        db.commit()
        return NaturalLanguageQueryResponse(
            supported=False,
            message="That question is outside the supported business analytics capabilities.",
        )

    try:
        filters = analytics_service.filters_for_question(payload.question, payload.filters)
        start, end, data = analytics_service.execute(
            db,
            membership.organization_id,
            intent,
            filters,
        )
        log_activity(
            db,
            organization_id=membership.organization_id,
            user=membership.user,
            action="ANALYTICS_NL_QUERY_EXECUTED",
            entity_type="ANALYTICS",
            details=f"Executed controlled natural-language intent: {intent}",
        )
        db.commit()
    except Exception as exc:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Analytics request could not be completed") from exc

    return NaturalLanguageQueryResponse(
        supported=True,
        message=f"Answered using the supported {intent.replace('_', ' ')} capability.",
        intent=intent,
        date_from=start,
        date_to=end,
        data=data,
    )
